internal/ethapi/api.go

```diff
@@ -1286,12 +1286,15 @@ func DoEstimateGas(ctx context.Context, b CeloBackend, args TransactionArgs, blo
 		State:      state,
 		ErrorRatio: estimateGasErrorRatio,
 	}
+
 	// Celo specific: get exchange rates if fee currency is specified
 	exchangeRates := emptyExchangeRates
 	if args.FeeCurrency != nil {
-		// It is debatable whether we should use Hash or ParentHash here. Usually,
-		// user would probably like the recent rates after the block, so we use Hash.
-		exchangeRates, err = b.GetExchangeRates(ctx, header.Hash())
+		feeBlockNum, err := rpc.BlockNumberOrHashEnsureHashOnly(ctx, b, blockNrOrHash, true, false)
+		if err != nil {
+			return 0, err
+		}
+		exchangeRates, err = b.GetExchangeRates(ctx, feeBlockNum)
```

Comment:
Yes you are right that it is safe to just pass `blockNrOrHash`. If the pending state can't be retrieved it will use latest as a fallback.
However there is the tiny difference, that since now the pending state **is** retrievable, we use the current pending state during block building for gas estimation and the *newly updated exchange rates* that already applied during the block building. This is inconsistent with what we usually define as the exchange rates for a block, namely "freezing" the state to the end-state of the previously sealed block.

At this place this might not be too critical, since this function is estimating the gas-limit and thus an inconsistent exchange rate only has an effect on the maximum allowance in the gas estimation and potential execution failure due to insufficient funds.

Reply:
Done in 8ba11a40c83c6d118474d.

I decided to not go with the `pending` block distinction and use the passed in block itself for all queries. `Pending` usually means "the state during block building" so using the most recent exchange rates seemed more appropriate.